src/vs/platform/contextkey/common/scanner.ts

```diff
@@ -0,0 +1,387 @@
+
+		switch (peek) {
+			case ')':
+			case '=':
+			case '!':
+			case '&':
+			case '|':
+			case '~':
+				return false;
+			case '<':
+			case '>':
+				return (this._peekNext() === '=') ? false : true; // so that we support `foo<=1` as `foo <= 1`, but also support `vim.use<C-r>` as a single KEY
+			case '_':
+			case '-':
+			case '.':
+			case '/':
+			case '\\':
+			case ':':
+			case '*':
+			case '?': // do we have to?
+			case '%':
+			case '+':
+			case '[':
+			case ']':
+			case '^':
+			case ',':
+			case '#':
+			case '@':
+			case ';':
+			case '"':
+				return true;
+			default:
+				if (peek.charCodeAt(0) > 127) { // handle unicode, eg Chinese hieroglyphs used in extensions
+					return true;
+				}
+				return false;
+		}
+	}
+
+	private _isAlphaNumeric(ch: string) {
+		return (ch >= 'a' && ch <= 'z') || (ch >= 'A' && ch <= 'Z') || (ch >= '0' && ch <= '9');
+	}
+
+	private _isWhitespace(ch: string) {
+		return ch === ' ' || ch === '\t' || ch === '\n' || ch === '\r' || ch === /* &nbsp */ '\u00A0';
+	}
+
+	// captures the lexeme without the leading and trailing '
+	private _quotedString(): Token {
+		while (this._peek() !== `'` && !this._isAtEnd()) { // TODO@ulugbekna: add support for escaping ' ?
+			this._advance();
+		}
+
+		if (this._isAtEnd()) {
+			return this._error();
+		}
+
+		// consume the closing '
+		this._advance();
+
+		return { type: TokenType.QuotedStr, lexeme: this._input.substring(this._start + 1, this._current - 1), offset: this._start + 1 };
+	}
+
+	/*
+	 * Lexing a regex expression: /.../[igsmyu]*
+	 * Based on https://github.com/microsoft/TypeScript/blob/9247ef115e617805983740ba795d7a8164babf89/src/compiler/scanner.ts#L2129-L2181
+	 *
+	 * Note that we want slashes within a regex to be escaped, e.g., /file:\\/\\/\\// should match `file:///`
+	 */
+	private _regex(): Token {
+		let p = this._current;
+
+		let inEscape = false;
+		let inCharacterClass = false;
+		while (true) {
+			if (p >= this._input.length) {
+				return this._error();
+			}
+
+			const ch = this._input.charCodeAt(p);
+
+			if (inEscape) { // parsing an escape character
+				inEscape = false;
+			} else if (ch === CharCode.Slash && !inCharacterClass) { // end of regex
+				p++;
+				break;
+			} else if (ch === CharCode.OpenSquareBracket) {
+				inCharacterClass = true;
+			} else if (ch === CharCode.Backslash) {
+				inEscape = true;
+			} else if (ch === CharCode.CloseSquareBracket) {
+				inCharacterClass = false;
+			}
+			p++;
+		}
+
+		// Consume flags
+		while (p < this._input.length && Scanner._regexFlags.has(this._input.charCodeAt(p))) {
+			p++;
+		}
+
+		this._current = p;
+
+		return this._token(TokenType.RegexStr, true);
+	}
+
+	// invariant: this must not fail if at end of `this._value`
+	private _eatWhitespace() {
+		Scanner._whitespace.lastIndex = this._current;
+		const match = Scanner._whitespace.exec(this._input);
+		if (match) {
+			this._current += match[0].length;
+		}
```

Comment:
I believe this method doesn't do anything, because `/^\s*` matches any input and always returns a first match of length 0. I tried commenting out this method and it looks that all tests still pass. I think we should add a test that asserts that parsing `" a"` (leading whitespace) works correctly.

Reply:
You're right, `_eatWhitespace` was broken completely. Things worked because `next()` fn also handled whitespace.